### PR TITLE
add SlimAdamW optimizer

### PR DESCRIPTION
**Pull request by adistomar**
This PR adds the SlimAdamW optimizer

**Notes:**
- Run `bash install_slimadamw.sh` to install SlimAdamW and its dependencies
- Set `cramming_use_slimadamw` flag to True to use SlimAdamW
- SlimAdamW
   - currently doesn't support fusing
   - takes in `model` as its second argument: `slimadamw.SlimAdamW(model.parameters(), model, lr=...)`

Currently, SlimAdamW will only be used if
```py
cramming_fuse_optim_backward = False
cramming_offload_optim_cpu = False
cramming_use_slimadamw = True
```
I've not added SlimAdamW to `configure_fused_optimizers` since it doesn't support fusing yet. Also, SlimAdamW won't work with `CPUOffloadOptimizer` right now because it also takes `model` as an argument, so I'm guessing a wrapper class is needed to make this work.

### CODE REVIEW

**Comment by psandovalsegura** [comment]
Thanks @adistomar, I'll give this a try in next few days. One question: what's the size of the optim state for SlimAdamW in terms of model weights W?

**Comment by siddharth9820** [comment]
4W bytes

**Comment by adistomar** [comment]
> Thanks @adistomar, I'll give this a try in next few days. One question: what's the size of the optim state for SlimAdamW in terms of model weights W?

SlimAdamW compresses the second-order moments by maintaining one scalar per row of the matrix for linear layers. So
1. params = W
2. grads = W
3. momentum = W
4. second-order moments = (1/2) * W (where W = M x N and M ~= N, which is true for most layers except the embedding and output layers) 

So in total, 3W + 0.5W = 3.5 W, which is a decrease of ~12.5%.
For momentum and second-order moments specifically, W + 0.5W = 1.5W, which is a decrease of 25%.

During training (64 GPUs), we have noticed total memory savings of up to 20% for Llama2-70B. AdamW uses 2 x M x N values for momentum and second-order moments, and SlimAdamW cuts this down to M x (N + 1). So as model width increases, we see more memory savings

64_gpu_memory_consumption

**Comment by psandovalsegura** [comment]
But AdamW uses 2W for optimizer state. I expected this optim to have under
2W?

**Comment by adistomar** [comment]
Yes it is under 2W if you exclude params and gradients. By optim state I thought you also meant params and gradients.

**Comment by adistomar** [comment]
Optim state is ~1.5W

**Comment by psandovalsegura** [comment]
To train a Llama-2-7b model from scratch on fineweb, you first need to get llama tokens of fineweb:
```
# Use the Llama tokenizer on FineWeb-10b
export HF_HOME='/fs/nexus-scratch/psando/fineweb-10b'
python data/fineweb-10b/prepare.py
```

Next:
- make sure you're signed in to wandb (for logging)
- modify the [train_llama-2-7b_fineweb.py](https://github.com/psandovalsegura/cRAMming/blob/llama-2-7b/config/train_llama-2-7b_fineweb.py) config with your huggingface `cache_dir` and the location of your llama tokens `data_dir`

Then, you should be able to run the following slurm script:
```
#!/bin/bash
#SBATCH --account=scavenger
#SBATCH --job-name=lr-6e-4-fineweb-llama-a100
#SBATCH --time=3-0:00:00
#SBATCH --partition=scavenger
#SBATCH --qos=scavenger
#SBATCH --ntasks=1
#SBATCH --gres=gpu:a100:1
#SBATCH --cpus-per-task=4
#SBATCH --mem=64G
#SBATCH --output=slurm-%j-%x.out
#SBATCH --signal=B:USR1@120
#SBATCH --mail-type=begin,end,fail         
#SBATCH --mail-user=psando@umd.edu

# turning off all the cramming flags that make training slow.
export TOKENIZERS_PARALLELISM=true
exec python train_fineweb.py config/train_llama-2-7b_fineweb.py \
                --out_dir="cRAMming-experiments-output" \
                --gradient_accumulation_steps=80 \
                --cramming_offload_optim_cpu=False \
                --cramming_offload_gradients_cpu=False \
                --cramming_activation_checkpointing=False \
                --eval_interval=12 \
                --log_interval=12 \
                --learning_rate=6e-4 \
                --wandb_run_name='llama-2-7b' \
                --init_from="scratch"
```

File: train.py [hunk 1]
```diff
@@ -96,6 +98,7 @@ def get_block_class_from_model(model: torch.nn.Module, block_class_name: str) ->
 cramming_offload_gradients_cpu = True
 cramming_activation_checkpointing = True
 cramming_fuse_optim_backward = False
+cramming_use_slimadamw = False
```
Comment (psandovalsegura) [inline train.py:101]: let's import slimadamw only if `cramming_use_slimadamw` is true

Comment (psandovalsegura) [reply train.py:101]: and we can also add a comment that `cramming_use_slimadamw` only works when `cramming_offload_optim_cpu` is false
